File: packages/client-php/src/Exception/StaleEpochException.php
<?php

declare(strict_types=1);

namespace DurableStreams\Exception;

class StaleEpochException extends DurableStreamException
{
    public function __construct(
        protected int $currentEpoch,
        ?\Throwable $previous = null,
    ) {
        parent::__construct(
            "Stale epoch, current server epoch is {$currentEpoch}",
            'STALE_EPOCH',
            403,
            [],
            $previous
        );
    }

    public function getCurrentEpoch(): int
    {
        return $this->currentEpoch;
    }
}

File: packages/durable-streams/packages/client-java/src/main/java/com/durablestreams/exception/StaleEpochException.java
package com.durablestreams.exception;

/**
 * Thrown when an idempotent producer uses a stale epoch (403).
 */
public class StaleEpochException extends DurableStreamException {
    private final long currentEpoch;

    public StaleEpochException(long currentEpoch) {
        super("Stale epoch: current epoch is " + currentEpoch, 403, "STALE_EPOCH");
        this.currentEpoch = currentEpoch;
    }

    public long getCurrentEpoch() {
        return currentEpoch;
    }
}
